src/main/java/org/mule/extension/sftp/internal/proxy/http/HttpClientConnector.java

```diff
@@ -0,0 +1,250 @@
+    negotiate = new NegotiateAuthentication(this);
+    availableAuthentications = new ArrayList<>(2);
+    availableAuthentications.add(negotiate);
+    availableAuthentications.add(basic);
+    clientAuthentications = availableAuthentications.iterator();
+  }
+
+  private void close() {
+    HttpAuthenticationHandler current = authenticator;
+    authenticator = null;
+    if (current != null) {
+      current.close();
+    }
+  }
+
+  @Override
+  public void sendClientProxyMetadata(ClientSession sshSession)
+      throws Exception {
+    init(sshSession);
+    IoSession session = sshSession.getIoSession();
+    session.addCloseFutureListener(f -> close());
+    StringBuilder msg = connect();
+    if ((proxyUser != null && !proxyUser.isEmpty())
+        || (proxyPassword != null && proxyPassword.length > 0)) {
+      authenticator = basic;
+      basic.setParams(null);
+      basic.start();
+      msg = authenticate(msg, basic.getToken());
+      clearPassword();
+      proxyUser = null;
+    }
+    ongoing = true;
+    try {
+      send(msg, session);
+    } catch (Exception e) {
+      ongoing = false;
+      throw e;
+    }
+  }
+
+  private void send(StringBuilder msg, IoSession session) throws Exception {
+    byte[] data = eol(msg).toString().getBytes(US_ASCII);
+    Buffer buffer = new ByteArrayBuffer(data.length, false);
+    buffer.putRawBytes(data);
+    session.writeBuffer(buffer).verify(getTimeout());
+  }
+
+  private StringBuilder connect() {
+    StringBuilder msg = new StringBuilder();
+    // Persistent connections are the default in HTTP 1.1 (see RFC 2616),
+    // but let's be explicit.
+    return msg.append(format(
+                             "CONNECT HOST %s:%s HTTP/1.1%nProxy-Connection: keep-alive%nConnection: keep-alive%n",
+                             // $NON-NLS-1$
+                             remoteAddress.getHostString(),
+                             Integer.toString(remoteAddress.getPort())));
+  }
+
+  private StringBuilder authenticate(StringBuilder msg, String token) {
+    msg.append(HTTP_HEADER_PROXY_AUTHORIZATION).append(' ').append(token);
+    return eol(msg);
+  }
+
+  private StringBuilder eol(StringBuilder msg) {
+    return msg.append('\r').append('\n');
+  }
+
+  public void messageReceived(IoSession session, Readable buffer) throws Exception {
+    try {
+      int length = buffer.available();
+      byte[] data = new byte[length];
+      buffer.getRawBytes(data, 0, length);
+      String[] reply = new String(data, US_ASCII)
+          .split("\r\n"); //$NON-NLS-1$
+      handleMessage(session, Arrays.asList(reply));
+    } catch (Exception e) {
+      if (authenticator != null) {
+        authenticator.close();
+        authenticator = null;
+      }
+      ongoing = false;
+      try {
+        setDone(false);
+      } catch (Exception inner) {
+        e.addSuppressed(inner);
+      }
+      throw e;
+    }
+  }
+
+  private void handleMessage(IoSession session, List<String> reply)
```

Comment:
Consider using a chain of responsibility pattern for this handling. Check [this](https://refactoring.guru/design-patterns/chain-of-responsibility)